File: LeetCode/103/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
  vector<vector<int>> zigzagLevelOrder(TreeNode *root) {

    if (root == nullptr)
      return {};

    vector<vector<int>> ans;
    queue<TreeNode *> q;
    q.push(root);
    bool flag = true;

    while (!q.empty()) {

      int sz = q.size();
      vector<int> levels;

      for (int i = 0; i < sz; ++i) {
        TreeNode *node = q.front();
        q.pop();
        levels.emplace_back(node->val);
        if (node->left) {
          q.push(node->left);
        }
        if (node->right) {
          q.push(node->right);
        }
      }

      if (flag) {
        ans.emplace_back(levels);
      } else {
        ans.emplace_back(levels.rbegin(), levels.rend());
      }

      flag = !flag;
    }

    return ans;
  }
};

File: leetcode/103/main.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func rev(s *[]int) {
	for i, j := 0, len(*s)-1; i < j; i, j = i+1, j-1 {
		(*s)[i], (*s)[j] = (*s)[j], (*s)[i]
	}
}

func zigzagLevelOrder(root *TreeNode) [][]int {

	if root == nil {
		return [][]int{}
	}

	ans := [][]int{}
	q := []*TreeNode{root}

	flag := false

	for len(q) != 0 {

		sz := len(q)
		levels := []int{}

		for i := 0; i < sz; i++ {

			node := q[0]
			q = q[1:]
			levels = append(levels, node.Val)

			if node.Left != nil {
				q = append(q, node.Left)
			}
			if node.Right != nil {
				q = append(q, node.Right)
			}

		}

		if flag {
			rev(&levels)
		}
		ans = append(ans, levels)
		flag = !flag
	}

	return ans
}
